feat(curriculum): add email masker lab (#55768)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-email-masker/66b205e6eacba4c4e54ea434.md

---
id: 66b205e6eacba4c4e54ea434
title: Build an Email Masker
challengeType: 14
dashedName: build-an-email-masker
---

# --description--

In this lab, you will mask the username part of an email address with asterisks. Masking is a term used to hide or replace sensitive information with asterisks or other characters.

For example, if the email address was `myEmail@email.com`, then the masked email address will be `m*******l@email.com`.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. Create a function named `maskEmail` that takes `email` as an argument.

> 2. Inside the function, you should create a variable called `maskedName` and assign it the result of masking the `email` along with concatenating the domain name.  Remember that you can use methods like `slice`, `repeat`, `indexOf` `slice` or even `replace` to help you. Return `maskedName` from the function.

Reviewer: I agree with Tom here about the maskedName.
We were using that earlier because this lab originally didn't have functions.
But now that is does, we don't need it any more

Author: Thank you for this suggestion, Tom and Jessica. 
We won't need that user story now. I have updated the PR.

---
id: 66b205e6eacba4c4e54ea434
title: Build an Email Masker
challengeType: 14
dashedName: lab-email-masker
---

> # --description--
> 
> **Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

Reviewer: We should provide a description with an example here so they better understand.

```suggestion
# --description--

In this lab, you will mask the username part of an email address with asterisks. Masking is a term used to hide or replace sensitive information with asterisks or other characters.

For example, if the email address was `myEmail@email.com`, then the masked email address will be `m*******l@email.com`.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
```

**User Stories:**

> 1. Declare a variable named `email` to store the email address you want to mask. 

Reviewer: I am reopening this because the user story doesn't match the hint.
it should say to assign to assign `apple.pie@example.com` to it

> 2. Split the given email address by `@` using `slice()` to get the username and domain name. 
> 3. Store the username and domain name in `userName` and `domain` variables respectively.
> 4. The first letter and last letter of the username shouldn't change. Replace the remaining characters with asterisks (*) to mask it. Store the masked username in a variable named `maskedName`.
> 5. Combine the `maskedName` and `domain` to get the masked email address and store it in a variable named `finalMaskedEmail`.

Reviewer: We should combine all of these user stories into just one.
Because all we should care about is testing the final result.
The specific implementation and details should be left to the camper. 

Something like this could work. Or something along these lines.

```suggestion
2. You should create a variable called `maskedName` and assign it the result of masking the `email`.  Remember that you can use methods like `slice`, `repeat`, `indexOf` `slice` or even `replace` to help you.
```

> 6. Print the `finalMaskedEmail` to the console.

Reviewer: ```suggestion
6. Log the `maskedName` variable to the console. 
```

> # --hints--

Reviewer: we should update the tests to match the outlines suggestions I mentioned earlier.

---
id: 66b205e6eacba4c4e54ea434
title: Build an Email Masker
challengeType: 14
dashedName: build-an-email-masker
---

# --description--

In this lab, you will mask the username part of an email address with asterisks. Masking is a term used to hide or replace sensitive information with asterisks or other characters.

For example, if the email address was `myEmail@email.com`, then the masked email address will be `m*******l@email.com`.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. Declare a variable named `email` to store the email address you want to mask. 
2. You should create a variable called `maskedName` and assign it the result of masking the `email`.  Remember that you can use methods like `slice`, `repeat`, `indexOf` `slice` or even `replace` to help you.
3. Log the `maskedName` variable to the console. 

# --hints--

You should have an `email` variable.

```js
assert.isDefined(email);
```

You should assign `apple.pie@example.com` to your `email` variable.

```js
assert.equal(email, "apple.pie@example.com");
```

> Your masked email address should look like this: `a*******e@example.com`.

Reviewer: we should add another user story to make sure they don't hardcode the result like Tom mentioned earlier.
I would suggest experimenting with `assert.notMatch`
